File: Gold이상문제_정리/세용액_2473.java
package Gold이상문제_정리;

import java.util.*;
import java.io.*;

public class 세용액_2473 {
	static long[] Arr;
	static long[] save;
	static long Min = Long.MAX_VALUE;
	static long p1, p2, p3;
	static boolean done;
	public static void main(String[] args) throws IOException {
		//-97 -20 -6 -2 6 23 98
		/*
		-97을 뽑고 나머지에 대해서 -97에 가장 가까운 값을 이분탐색해서 찾는다
		 */
		BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
		StringTokenizer st = new StringTokenizer(br.readLine());

		int N = Integer.parseInt(st.nextToken());

		Arr = new long[N];
		st = new StringTokenizer(br.readLine());
		for (int i = 0; i < N; i++)
			Arr[i] = Long.parseLong(st.nextToken());

		//-97 -6 -2 6 98

		Arrays.sort(Arr);

		for (int i = 0; i < N - 1; i++) {
			save = Arrays.copyOfRange(Arr, i+1, Arr.length);
			find(0, save.length - 1, Arr[i]);
			if (done)
				break;
		}

		System.out.println(p1 + " " + p2 + " " + p3);
	}
	private static void find(int start, int end, long key)
	{
		while (start < end)
		{
			long calc = save[start] + save[end] + key;
			// 0보다 작은 경우 더 증가
			if (calc < 0)
			{
				if (Math.abs(calc) < Min) {
					Min = Math.abs(calc);
					p1 = key;
					p2 = save[start];
					p3 = save[end];
				}
				start++;
			}
			else if (calc > 0)
			{
				if (Math.abs(calc) < Min) {
					Min = Math.abs(calc);
					p1 = key;
					p2 = save[start];
					p3 = save[end];
				}
				end--;
			}
			else {
				done = true;
				Min = 0;
				p1 = key;
				p2 = save[start];
				p3 = save[end];
				return;
			}
		}
	}
}

File: 코틀린_코테스터디/week4/세용액_2473.kt
import java.io.BufferedReader
import java.io.InputStreamReader
import java.util.*
import kotlin.math.abs

/*
7
-2 -3 -24 -6 98 100 61

-24 -6 -3 -2 61 98 100

-97 -6 -2 6 98
 */

private lateinit var liquids: Array<Long>
private lateinit var subLiquids: Array<Long>
private var N = 0
private var MIN = Long.MAX_VALUE
private var abc: Triple<Long, Long, Long> = Triple(0, 0, 0)
private var noMore: Boolean = false

fun main() {
    val br = BufferedReader(InputStreamReader(System.`in`))
    N = br.readLine().toInt()
    liquids = Array(N) { 0 }

    val st = StringTokenizer(br.readLine())
    repeat(N) { i ->
        liquids[i] = st.nextToken().toLong()
    }

    Arrays.sort(liquids)

    for (index in 0 until N) {
        subLiquids = liquids.copyOfRange(index + 1, N)
        binarySearch(0, subLiquids.size - 1, liquids[index])
        if (noMore) {
            print("${abc.first} ${abc.second} ${abc.third}")
            return
        }
    }
    print("${abc.first} ${abc.second} ${abc.third}")
}

fun binarySearch(start: Int, end: Int, key: Long) {
    var sum: Long = 0
    var start = start
    var end = end

    while (start < end) {
        sum = subLiquids[start] + subLiquids[end] + key

        if (sum > 0) {
            if (abs(sum) < MIN) {
                MIN = abs(sum)
                abc = Triple(key, subLiquids[start], subLiquids[end])
            }
            end--
        }
        else if (sum < 0) {
            if (abs(sum) < MIN) {
                MIN = abs(sum)
                abc = Triple(key, subLiquids[start], subLiquids[end])
            }
            start++
        }
        else {
            noMore = true
            abc = Triple(key, subLiquids[start], subLiquids[end])
            break
        }
    }
}
